File: Chroma-Core/src/main/java/buttondevteam/core/PlayerListener.java
package buttondevteam.core;

import buttondevteam.core.component.towny.TownyComponent;
import buttondevteam.lib.*;
import buttondevteam.lib.architecture.ButtonPlugin;
import buttondevteam.lib.chat.ChatMessage;
import buttondevteam.lib.chat.Command2MCSender;
import buttondevteam.lib.chat.TBMCChatAPI;
import buttondevteam.lib.player.ChromaGamerBase;
import buttondevteam.lib.player.TBMCPlayer;
import buttondevteam.lib.player.TBMCPlayerBase;
import lombok.val;
import org.bukkit.Bukkit;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.event.Cancellable;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.server.ServerCommandEvent;

import java.util.Arrays;

public class PlayerListener implements Listener {

	@EventHandler(priority = EventPriority.NORMAL)
	public void OnPlayerJoin(PlayerJoinEvent event) {
		var p = event.getPlayer();
		TBMCPlayer player = TBMCPlayerBase.getPlayer(p.getUniqueId(), TBMCPlayer.class);
		String pname = player.PlayerName.get();
		if (pname.length() == 0) {
			player.PlayerName.set(p.getName());
			MainPlugin.Instance.getLogger().info("Player name saved: " + player.PlayerName.get());
		} else if (!p.getName().equals(pname)) {
			TownyComponent.renameInTowny(pname, p.getName());
			MainPlugin.Instance.getLogger().info(pname + " renamed to " + p.getName());
			player.PlayerName.set(p.getName());
		}
	}

	@EventHandler(priority = EventPriority.NORMAL)
	public void OnPlayerLeave(PlayerQuitEvent event) {
		TBMCPlayerBase.getPlayer(event.getPlayer().getUniqueId(), TBMCPlayer.class).uncache();
	}

	@EventHandler(priority = EventPriority.HIGHEST)
	public void onSystemChat(TBMCSystemChatEvent event) {
		if (event.isHandled())
			return; // Only handle here if ButtonChat couldn't - ButtonChat doesn't even handle this
		if (Arrays.stream(event.getExceptions()).anyMatch("Minecraft"::equalsIgnoreCase))
			return;
		Bukkit.getOnlinePlayers().stream().filter(event::shouldSendTo)
			.forEach(p -> p.sendMessage(event.getChannel().DisplayName.get().substring(0, 2) + event.getMessage()));
	}

	@EventHandler
	public void onPlayerChatPreprocess(PlayerCommandPreprocessEvent event) {
		handlePreprocess(event.getPlayer(), event.getMessage(), event);
	}

	@EventHandler
	public void onSystemChatPreprocess(ServerCommandEvent event) {
		handlePreprocess(event.getSender(), "/" + event.getCommand(), event);
		if (event.isCancelled()) event.setCommand("dontrunthiscmd"); //Bugfix
	}

	private void handlePreprocess(CommandSender sender, String message, Cancellable event) {
		if (event.isCancelled()) return;
		val cg = ChromaGamerBase.getFromSender(sender);
		if (cg == null) throw new RuntimeException("Couldn't get user from sender for " + sender.getName() + "!");
		val ev = new TBMCCommandPreprocessEvent(sender, cg.channel.get(), message, sender);
		Bukkit.getPluginManager().callEvent(ev);
		if (ev.isCancelled())
			event.setCancelled(true); //Cancel the original event
	}

	@EventHandler
	public void onTBMCPreprocess(TBMCCommandPreprocessEvent event) {
		if (event.isCancelled()) return;
		try {
			event.setCancelled(ButtonPlugin.getCommand2MC().handleCommand(new Command2MCSender(event.getSender(), event.getChannel(), event.getPermCheck()), event.getMessage()));
		} catch (Exception e) {
			TBMCCoreAPI.SendException("Command processing failed for sender '" + event.getSender() + "' and message '" + event.getMessage() + "'", e, MainPlugin.Instance);
		}
	}

	@EventHandler(priority = EventPriority.HIGH) //The one in the chat plugin is set to highest
	public void onPlayerChat(AsyncPlayerChatEvent event) {
		if (event.isCancelled())
			return; //The chat plugin should cancel it after this handler
		val cp = TBMCPlayer.getPlayer(event.getPlayer().getUniqueId(), TBMCPlayer.class);
		TBMCChatAPI.SendChatMessage(ChatMessage.builder(event.getPlayer(), cp, event.getMessage()).build());
		//Not cancelling the original event here, it's cancelled in the chat plugin
		//This way other plugins can deal with the MC formatting if the chat plugin isn't present, but other platforms still get the message
	}

	@EventHandler(priority = EventPriority.HIGH) //The one in the chat plugin is set to highest
	public void onPlayerChat(TBMCChatEvent event) {
		if (event.isCancelled())
			return;
		if (!MainPlugin.Instance.isChatHandlerEnabled()) return;
		if (event.getOrigin().equals("Minecraft")) return; //Let other plugins handle MC messages
		var channel = event.getChannel();
		String msg = MainPlugin.Instance.chatFormat.get()
			.replace("{channel}", channel.DisplayName.get())
			.replace("{origin}", event.getOrigin().substring(0, 1))
			.replace("{name}", ChromaUtils.getDisplayName(event.getSender()))
			.replace("{message}", String.format("§%x%s", channel.Color.get().ordinal(), event.getMessage()));
		for (Player player : Bukkit.getOnlinePlayers())
			if (event.shouldSendTo(player))
				player.sendMessage(msg);
		Bukkit.getConsoleSender().sendMessage(msg);
	}
}

File: Chroma-Core/src/main/java/buttondevteam/core/PlayerListener.kt
package buttondevteam.core

import buttondevteam.lib.TBMCChatEvent
import buttondevteam.lib.TBMCCommandPreprocessEvent
import buttondevteam.lib.TBMCCoreAPI
import buttondevteam.lib.TBMCSystemChatEvent
import buttondevteam.lib.architecture.ButtonPlugin
import buttondevteam.lib.chat.ChatMessage
import buttondevteam.lib.chat.Command2MCSender
import buttondevteam.lib.chat.TBMCChatAPI
import buttondevteam.lib.player.ChromaGamerBase
import buttondevteam.lib.player.TBMCPlayer
import buttondevteam.lib.player.TBMCPlayerBase
import buttondevteam.lib.player.TBMCPlayerBase.Companion.asTBMC
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import org.bukkit.event.Cancellable
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.event.player.PlayerCommandPreprocessEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.event.server.ServerCommandEvent

class PlayerListener(val plugin: MainPlugin) : Listener {
    @EventHandler(priority = EventPriority.NORMAL)
    fun onPlayerJoin(event: PlayerJoinEvent) {
        val p = event.player
        val player = TBMCPlayerBase.getPlayer(p.uniqueId, TBMCPlayer::class.java)
        val pname = player.playerName
        if (pname.isEmpty()) {
            player.playerName = p.name
            plugin.logger.info("Player name saved: " + player.playerName)
        } else if (p.name != pname) {
            plugin.logger.info(pname + " renamed to " + p.name)
            player.playerName = p.name
        }
    }

    @EventHandler(priority = EventPriority.NORMAL)
    fun onPlayerLeave(event: PlayerQuitEvent) {
        TBMCPlayerBase.getPlayer(event.player.uniqueId, TBMCPlayer::class.java).uncache()
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onSystemChat(event: TBMCSystemChatEvent) {
        if (event.isHandled) return
        if (event.exceptions.any { "Minecraft".equals(it, ignoreCase = true) }) return
        Bukkit.getOnlinePlayers().stream().filter { event.shouldSendTo(it.asTBMC()) }
            .forEach { it.sendMessage(event.channel.displayName.get().substring(0, 2) + event.message) }
        // TODO: Replace the substring call with a proper color code getter.
    }

    @EventHandler
    fun onPlayerChatPreprocess(event: PlayerCommandPreprocessEvent) {
        handlePreprocess(event.player, event.message, event)
    }

    @EventHandler
    fun onSystemChatPreprocess(event: ServerCommandEvent) {
        handlePreprocess(event.sender, "/" + event.command, event)
        if (event.isCancelled) event.command = "dontrunthiscmd" //Bugfix
    }

    private fun handlePreprocess(sender: CommandSender, message: String, event: Cancellable) {
        if (event.isCancelled) return
        val cg = ChromaGamerBase.getFromSender(sender)
        val ev = TBMCCommandPreprocessEvent(cg, cg.channel.get(), message, cg)
        Bukkit.getPluginManager().callEvent(ev)
        if (ev.isCancelled) event.isCancelled = true //Cancel the original event
    }

    @EventHandler
    fun onTBMCPreprocess(event: TBMCCommandPreprocessEvent) {
        if (event.isCancelled) return
        try {
            val sender = Command2MCSender(event.sender, event.channel, event.permCheck)
            event.isCancelled = ButtonPlugin.command2MC.handleCommand(sender, event.message)
        } catch (e: Exception) {
            TBMCCoreAPI.SendException(
                "Command processing failed for sender '${event.sender}' and message '${event.message}'",
                e,
                plugin
            )
        }
    }

    @EventHandler(priority = EventPriority.HIGH) //The one in the chat plugin is set to highest
    fun onPlayerChat(event: AsyncPlayerChatEvent) {
        if (event.isCancelled) return  //The chat plugin should cancel it after this handler
        val cp = TBMCPlayerBase.getPlayer(event.player.uniqueId, TBMCPlayer::class.java)
        TBMCChatAPI.sendChatMessage(ChatMessage.builder(cp, event.message).build())
        //Not cancelling the original event here, it's cancelled in the chat plugin
        //This way other plugins can deal with the MC formatting if the chat plugin isn't present, but other platforms still get the message
    }

    @EventHandler(priority = EventPriority.HIGH) //The one in the chat plugin is set to highest
    fun onPlayerChat(event: TBMCChatEvent) {
        if (event.isCancelled) return
        if (!plugin.isChatHandlerEnabled) return
        if (event.origin == "Minecraft") return  //Let other plugins handle MC messages
        val channel = event.channel
        val msg = plugin.chatFormat.get()
            .replace("{channel}", channel.displayName.get())
            .replace("{origin}", event.origin.substring(0, 1))
            .replace("{name}", event.user.name)
            .replace("{message}", String.format("§%x%s", channel.color.get().ordinal, event.message))
        for (player in Bukkit.getOnlinePlayers()) if (event.shouldSendTo(player.asTBMC())) player.sendMessage(msg)
        Bukkit.getConsoleSender().sendMessage(msg)
    }
}